Gemfile

```diff
@@ -7,6 +7,8 @@ end
 
 
 gem "rails", "~> 5.1.5"
+gem "config"
+gem "bcrypt", "3.1.11"
```

Comment:
gem này để làm gì vậy em?

Reply:
gem config dùng để quản lý và sử dụng các settings trong project ạ.
gem bcrypt là dùng để mã hóa password

Reply:
cơ chế mã hóa là gì?

Reply:
Sử dụng hàm băm để mã hóa ạ

Reply:
cơ chế băm như thế nào hả em?

Reply:
Cơ chế băm: với một chuỗi đầu vào thì hàm băm luôn cho ra một chuỗi có độ dài cố định định trước.

Reply:
có nghĩa là với cùng 1 chuỗi đầu vào, mọi lần băm đều như nhau?

Reply:
Với một chuỗi băm đầu vào giống nhau thì mọi lần băm sẽ cho ra chuỗi băm giống nhau( trừ một số thuật toán có cộng thêm chuỗi salt thì kết quả băm sẽ không giống nhau)

Reply:
không hiểu lắm

Reply:
Với chuỗi băm đầu vào giống nhau thì sau khi băm sẽ cho ra chuỗi băm giống nhau. Còn về chuỗi salt là chuỗi được thêm vào sau mỗi lần người dùng nhập chuỗi, khi mã hóa thì sẽ mã hóa chuỗi do người dùng nhập vào + chuỗi salt( cơ chế này thường áp dụng trong linux khi tạo user mới)

Reply:
chưa đúng rồi em ơi, đọc thêm đi nhé

Reply:
Chuỗi băm không giống nhau =]]]

Reply:
Chuỗi băm không giống nhau với đầu vào giống nhau

Reply:
vậy sao biết là đó là mật khẩu đúng?

Reply:
Từ chuỗi salt sinh ra ở trên thì BCrypt thực hiện mã hóa mật khẩu để thu được một chuỗi có định dạng như chuỗi: $2a$10$Ww/nuq6UNXXIWaUaAg.FDe4vHKwTdAK4cqldlNDtyj63M5Tv034xS.
Trong chuỗi ví trên được chia làm 3 phần được phân cách bởi dấu "$". 
bcryt sử dụng hàm hash_secret, dựa vào chuỗi salt và mật khẩu người dùng nhập để sinh ra chuỗi hash giống chuỗi hash đã lưu trong database

Reply:
3 phần cụ thể là những phần nào? tác dụng của từng phần?